module/zfs/dsl_scan.c

```diff
@@ -2386,6 +2386,20 @@ dsl_scan_ddt_entry(dsl_scan_t *scn, enum zio_checksum checksum,
 	if (!dsl_scan_is_running(scn))
 		return;
 
+	/*
+	 * This function is special because it is the only thing
+	 * that can add scan_io_t's to the vdev scan queues from
+	 * outside dsl_scan_sync(). For the most part this is ok
+	 * as long as it is called from within syncing context.
+	 * However, dsl_scan_sync() expects that no new sio's will
+	 * be added between when all the work for a scan is done
+	 * and the next txg when the scan is actually marked as
+	 * completed. This check ensures we do not sio's during
+	 * this period.
+	 */
+	if (scn->scn_done_txg != 0 && scn->scn_bytes_pending == 0)
```

Comment:
This is to match the check in `dsl_scan_sync()`. We don't necessarily need that checkdepending on what we want. Having it there basically just ensures that if the scan is running we still scan incoming dedup blocks, even if the scan if finishing. Note that the "finishing" part of scanning can still take several  hours depending on the block sizes and RAM given to the dsl scan queues.

Reply:
Which check in dsl_scan_sync()?  And what exactly do you mean by "finishing"?  Once scn_done_txg is set, could we still wait hours?

Reply:
> Which check in dsl_scan_sync()?
```
else if (scn->scn_is_sorted && scn->scn_bytes_pending != 0) {
```

> And what exactly do you mean by "finishing"? Once scn_done_txg is set, could we still wait hours?

Yes. `scn_done_txg` is set when the last block of the scan is placed on one of the per-vdev scanning queues. The scan is not complete, however until all of the queued I/O has been issued. Back when I was testing this code, I saw this take about an hour on my test server, but I imagine it could be longer with other configurations. This check basically says "we won't add anything more to the queues after the normal scanning has completed adding its I/Os to the queues".

Reply:
I see.  I wasn't that familiar with the changes made due to sorted scan here.  Once `scn_done_txg` is set, we've found all the blocks to scan (presumably including dedup blocks).  Once that has happened, do we still need to do explicit scans in dsl_scan_ddt_entry()?  I don't think so.

Reply:
Ok. ill make the change.